fix: Enable num_return_sequences (`n`) support in PyTorch backend

jaedeok-nvidia opened the pull request:
This PR enables the `n` parameter (`num_return_sequences`) in the PyTorch backend, which is the default path for LLM API. While this feature was already implemented in the TRT backend via C++ Executor, it was missing in the PyExecutor. This PR fixes the gap by adding necessary APIs to the pybind of the LlmRequest class.

**Changes**:
  - Added `create_child_request` method to pyexecutor.LlmRequest that wraps C++'s createChildRequest method. This allows requests to properly handle their child requests and states.
  - Updated C++ LlmRequest and related Python bindings to expose additional properties required in the PyTorch backend.
  - Enhanced PyExecutor to create child requests, ensuring proper handling of requests when num_return_sequences > 1.

jaedeok-nvidia committed:
fix: Enable num_return_sequences (`n`) support in PyTorch backend

This PR enables the `n` parameter (num_return_sequences) in the PyTorch backend, which is the default path for LLM API. While this feature was already implemented in the TRT backend via C++ Executor, it was missing in the PyExecutor. This PR fixes the gap by adding necessary APIs to the pybind of the `LlmRequest` class.

Changes:
  - Added `create_child_request` method to `pyexecutor.LlmRequest` that wraps C++'s createChildRequest method. This allows requests to properly handle their child requests and states.
  - Updated C++ LlmRequest and related Python bindings to expose additional properties required in the PyTorch backend.
  - Enhanced `PyExecutor` to create child requests, ensuring proper handling of requests when `num_return_sequences > 1`.

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia committed:
fix child request's  create_response

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia committed:
add `n` to pytorch example

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia committed:
add llm request unittest

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

syuoni commented on tensorrt_llm/_torch/pyexecutor/llm_request.py:
```diff
@@ -206,19 +207,15 @@ class LlmResult:
     def __getattr__(self, item):
         if item in self.py_result_properties:
             return getattr(self._py_result, item)
-        if item == 'is_final':
-            return object.__getattribute__(self, 'is_final')
```
I think the `is_final` at this level is to keep the attribute access in Python, which has lower overhead than attribute access  from Pybind objects.

We do this for those frequently accessed attributes. cc @QiJune

syuoni commented on tensorrt_llm/_torch/pyexecutor/llm_request.py:
```diff
@@ -297,19 +313,71 @@ def __init__(
+        # Create a child request by C++'s API to track the states each other.
+        child_request = super().create_child_request(request_id)
+
+        # Copy Python-specific attributes from parent to child
+        child_request.py_client_id = self.py_client_id
+        child_request.py_parent_request_id = self.py_request_id
+        child_request.py_request_id = child_request.request_id
+        child_request.py_llm_request_type = child_request.llm_request_type
+        child_request.py_end_id = child_request.end_id
+        child_request.py_prompt_len = child_request.prompt_len
+        child_request.py_orig_prompt_len = child_request.orig_prompt_len
+        child_request.py_max_new_tokens = child_request.max_new_tokens
+
+        # Copy Python-specific configuration from parent
+        child_request.py_return_log_probs = self.py_return_log_probs
+        child_request.py_return_context_logits = self.py_return_context_logits
+        child_request.py_return_generation_logits = self.py_return_generation_logits
+        child_request.py_return_logits_device_memory = self.py_return_logits_device_memory
+        child_request.py_exclude_last_generation_logits = self.py_exclude_last_generation_logits
+        child_request.py_stop_words_list = self.py_stop_words_list
+        child_request.py_logits_post_processors = self.py_logits_post_processors
+        child_request.py_rewind_len = self.py_rewind_len
+        child_request.py_decoding_iter = self.py_decoding_iter
+        child_request.py_draft_tokens = self.py_draft_tokens.copy(
+        ) if self.py_draft_tokens else []
+        child_request.py_last_draft_tokens = self.py_last_draft_tokens.copy(
+        ) if self.py_last_draft_tokens else None
+        child_request.py_num_accepted_draft_tokens = self.py_num_accepted_draft_tokens
+        child_request.py_lora_task_layer_module_configs = self.py_lora_task_layer_module_configs
+
+        # Initialize Python-specific runtime state
+        child_request.py_batch_idx = None
+        child_request.is_attention_dp_dummy = self.is_attention_dp_dummy
+        child_request.is_cuda_graph_dummy = self.is_cuda_graph_dummy
```
If possible, we should make this happen automatically instead of manually copying every field. Otherwise, we need to maintain this list every time when adding or removing an attribute.

Will a `copy.deepcopy` work?

Superjomn replied:
I see. But considering there are so many members flattened, and it could be easy to forget when new member is introduced. Maybe the following code can help automate the copying of most of the members:
```python
for attr_name in dir(self):
    if attr_name.startswith("py_"):
        value = getattr(self, attr_name)
        setattr(child_request, attr_name, value)
```

You can try it in a subsequent PR, not necessary to change in this PR.

jaedeok-nvidia replied:
```python
        # Copy all py_* attributes from parent to child
        for attr_name, attr_value in self.__dict__.items():
            if attr_name.startswith('py_'):
                attr_value = getattr(self, attr_name)
                setattr(child_request, attr_name, copy.deepcopy(attr_value))
```
Agree it's easy to forget as there are many contributors. However, it's already updated tho, github shows the original impl not a revision.

syuoni replied:
I see. Currently there is no method to create child request as `pyexecutor.LlmRequest`.

Are those children requests be processed in Python runtime? If so, will its different type (`bindings.LlmRequest`) cause any issues?

jaedeok-nvidia replied:
@syuoni  yes, that's why `mimic` functions are added. however, as mentioned, it is just a WAR before properly reimplementing the `LlmRequest` and Resource manager logics within torch backend.

syuoni commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -428,15 +460,34 @@ def enqueue_request(self,
+            # Only create child requests if beam_width == 1 and num_return_sequences > 1
+            if beam_width == 1 and num_return_sequences > 1:
+                child_req_ids = []
+                for i in range(num_return_sequences - 1):
```
Is the first request also the parent request?

jaedeok-nvidia replied:
Maybe I misunderstand the question tho. This block will reserve `(n - 1)` request ids from the parent id and RequestQueue has a dedicated field for keeping `child_req_ids`.

jaedeok-nvidia committed:
simplify create_child_request logic

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia committed:
fix format

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia commented:
/bot run

ccys-a11y commented:
@jaedeok-nvidia   I reinstall the tensorrt_llm under your branch "https://github.com/jaedeok-nvidia/TensorRT-LLM.git", the problem still exists when n=2.  Error Info: "Processed requests:   0%|                                                                                                  | 0/4 [00:00<?, ?it/s][07/08/2025-17:05:26] [TRT-LLM] [E] Error in event loop: fail to schedule any pending request, probably run out of resource.
[07/08/2025-17:05:26] [TRT-LLM] [E] Traceback (most recent call last):
  File "/root/new/TensorRT-LLM/tensorrt_llm/_torch/pyexecutor/py_executor.py", line 285, in _event_loop_wrapper
    self.event_loop()
  File "/root/new/TensorRT-LLM/tensorrt_llm/_torch/pyexecutor/py_executor.py", line 1033, in _executor_loop_overlap
    assert scheduled_batch.batch_size > 0, (
AssertionError: fail to schedule any pending request, probably run out of resource.

Exception in thread Thread-7 (_event_loop_wrapper):
Traceback (most recent call last):
  File "/usr/local/python3/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/local/python3/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/root/new/TensorRT-LLM/tensorrt_llm/_torch/pyexecutor/py_executor.py", line 289, in _event_loop_wrapper
    raise e
  File "/root/new/TensorRT-LLM/tensorrt_llm/_torch/pyexecutor/py_executor.py", line 285, in _event_loop_wrapper
    self.event_loop()
  File "/root/new/TensorRT-LLM/tensorrt_llm/_torch/pyexecutor/py_executor.py", line 1033, in _executor_loop_overlap
    assert scheduled_batch.batch_size > 0, (
AssertionError: fail to schedule any pending request, probably run out of resource."

jaedeok-nvidia committed:
fix py llm request

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia commented:
Hi @ccys-a11y sorry for the inconvenience. This PR was broken while rebasing the main and addressing review comments. For now, I confirmed it works at least `quickstart_advanced.py` and am on further tests. Here is the command that I used for a quick test (`TinyLlama-1.1B-Chat-v1.0` was used.).

```bash
# Two sequences should be identical due to greedy decoding.
$ TLLM_ALLOW_N_GREEDY_DECODING=1 python quickstart_advanced.py --model_dir /path/to/model --n 2
# Two sequences are expected to be different since high temperature makes almost random.
$ python quickstart_advanced.py --model_dir /path/to/model --n 2 --top_p 0.9 --temperature 999
```

For the second issue (2. Failed to import LLM from torch: "from tensorrt_llm._torch import LLM”), I think this is not directly related to this PR. We've made torch backend as the default path few weeks ago. That was the reason I guess. You can directly import LLM by `from tensorrt_llm import LLM`. However, for the current branch after rebasing ToT, it seems working.
```bash
>>> from tensorrt_llm._torch import LLM
...
[07/09/2025-06:47:24] [TRT-LLM] [I] Starting TensorRT-LLM init.
[TensorRT-LLM][INFO] Set logger level to INFO
2025-07-09 06:47:24,959 - INFO - flashinfer.jit: Prebuilt kernels not found, using JIT backend
[07/09/2025-06:47:25] [TRT-LLM] [I] TensorRT-LLM inited.
[TensorRT-LLM] TensorRT-LLM version: 1.0.0rc3
```

tensorrt-cicd commented:
[PR_Github #11404](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11404/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8435](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8435/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #11428](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11428/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11428](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11428/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #8452](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8452/) completed with status: 'FAILURE'

jaedeok-nvidia committed:
fix llm request unittest

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia commented:
/bot run --disable-fail-fast

jaedeok-nvidia committed:
rollback result serialization to reduce host overhead

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia commented:
/bot run

tensorrt-cicd commented:
[PR_Github #11512](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11512/) [ run ] triggered by Bot

jaedeok-nvidia committed:
remove unnecessary api expose

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia commented:
/bot kill

jaedeok-nvidia commented:
/bot run

jaedeok-nvidia committed:
fix eagle case

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

tensorrt-cicd commented:
[PR_Github #11574](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11574/) [ run ] triggered by Bot

syuoni reviewed (approved):
No major concerns from my side. Thanks!

Superjomn reviewed (approved):
LGTM

QiJune commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -438,15 +470,34 @@ def enqueue_request(self,
-                self.request_queue.put(RequestQueueItem(req_id, request))
-            self.next_req_id += 1
+            # Generate child request IDs if needed
+            child_req_ids = None
```
It seems that line480 to line 500 shares the same codes with line 343 to line 361, can we provide a helper function?

jaedeok-nvidia committed:
factor out child req id generation logic

Signed-off-by: Jaedeok Kim <jaedeokk@nvidia.com>

jaedeok-nvidia commented on tensorrt_llm/_torch/pyexecutor/llm_request.py:
```diff
@@ -327,28 +363,64 @@ def __init__(
+    def is_generation_only(self):
+        # is_generation_only_request is a property getter at the C++ side,
+        # so here use a different name at the pytorch backend.
+        return is_generation_only_request(self)
```
Thanks for the discussion. Reducing the host overhead is important as many LlmRequest instances are called here and there. 

As explained offline, regarding `is_generation_only_request` function it's a readonly property at c++ binding, so that it cannot be overwritten at the child request. Since currently a child request of type batch_manager.LlmRequest not a py LlmRequest. In order to work seamlessly, it is needed to overwrite the function. That's why I rename it. Instead, when calling the function, it is now calling this python function instead of binding one. This should be gone after re-impl completes.

tensorrt-cicd commented:
[PR_Github #11654](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11654/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8630](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8630/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #11731](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11731/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8686](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8686/) completed with status: 'FAILURE'

QiJune reviewed (approved):
LGTM

tensorrt-cicd commented:
[PR_Github #11748](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11748/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8700](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8700/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/8700/test-results/rerun_report.html) for details.

ccys-a11y commented:
Hi@jaedeok-nvidia，thanks for your method. I find it works for 'quickstart' script. However, when I benchmark Qwen3-14B for AIME24/25 datasets with n=32, the following error might occur intermittently. It seems it's not stable enough. Can you help? 

"
Traceback (most recent call last):
  File "/usr/local/python3/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/local/python3/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/python3/lib/python3.10/site-packages/tensorrt_llm/_torch/pyexecutor/py_executor.py", line 1031, in _executor_loop_overlap
    self.resource_manager.prepare_resources(scheduled_batch)
  File "/usr/local/python3/lib/python3.10/site-packages/tensorrt_llm/_torch/pyexecutor/resource_manager.py", line 793, in prepare_resources
    resource_manager.prepare_resources(scheduled_batch)
  File "/usr/local/python3/lib/python3.10/site-packages/tensorrt_llm/_torch/pyexecutor/resource_manager.py", line 307, in prepare_resources
    self.impl.add_token(req.py_request_id)
RuntimeError: [TensorRT-LLM][ERROR] Assertion failed: isLeaf() (/dockerdata/caiyi/tensorrt_llm/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:334)
1       0x7f8b7f48b9a9 tensorrt_llm::common::throwRuntimeError(char const*, int, char const*) + 76
2       0x7f8b7f4b75c2 /usr/local/python3/lib/python3.10/site-packages/tensorrt_llm/libs/libtensorrt_llm.so(+0x1a615c2) [0x7f8b7f4b75c2]
3       0x7f8b801d2380 tensorrt_llm::batch_manager::kv_cache_manager::WindowBlockManager::getFreeBlock(int, std::optional<std::chrono::duration<long, std::ratio<1l, 1000l> > >) + 176
4       0x7f8b801d3f7b tensorrt_llm::batch_manager::kv_cache_manager::WindowBlockManager::allocateBlock(tensorrt_llm::batch_manager::kv_cache_manager::GenerationRequest&, bool) + 299
5       0x7f8b801d560b tensorrt_llm::batch_manager::kv_cache_manager::KVCacheManager::updateToken(tensorrt_llm::batch_manager::kv_cache_manager::GenerationRequest&, bool) + 123
6       0x7f8b95ad7275 /usr/local/python3/lib/python3.10/site-packages/tensorrt_llm/bindings.cpython-310-x86_64-linux-gnu.so(+0x108275) [0x7f8b95ad7275]
7       0x7f8b95a7e040 /usr/local/python3/lib/python3.10/site-packages/tensorrt_llm/bindings.cpython-310-x86_64-linux-gnu.so(+0xaf040) [0x7f8b95a7e040]
8       0x7f8dc201284e /usr/local/python3/lib/libpython3.10.so.1.0(+0x1f384e) [0x7f8dc201284e]
9       0x7f8dc1fecedb _PyObject_MakeTpCall + 123
10      0x7f8dc1f1f27d /usr/local/python3/lib/libpython3.10.so.1.0(+0x10027d) [0x7f8dc1f1f27d]
11      0x7f8dc2051dc7 _PyEval_EvalFrameDefault + 20487
12      0x7f8dc204bfe2 /usr/local/python3/lib/libpython3.10.so.1.0(+0x22cfe2) [0x7f8dc204bfe2]
13      0x7f8dc204d49a _PyEval_EvalFrameDefault + 1754
14      0x7f8dc204bfe2 /usr/local/python3/lib/libpython3.10.so.1.0(+0x22cfe2) [0x7f8dc204bfe2]
15      0x7f8dc204d49a _PyEval_EvalFrameDefault + 1754
16      0x7f8dc204bfe2 /usr/local/python3/lib/libpython3.10.so.1.0(+0x22cfe2) [0x7f8dc204bfe2]
17      0x7f8dc1fef1ee /usr/local/python3/lib/libpython3.10.so.1.0(+0x1d01ee) [0x7f8dc1fef1ee]
18      0x7f8dc204fc90 _PyEval_EvalFrameDefault + 11984
19      0x7f8dc204bfe2 /usr/local/python3/lib/libpython3.10.so.1.0(+0x22cfe2) [0x7f8dc204bfe2]
20      0x7f8dc204d49a _PyEval_EvalFrameDefault + 1754
21      0x7f8dc204bfe2 /usr/local/python3/lib/libpython3.10.so.1.0(+0x22cfe2) [0x7f8dc204bfe2]
22      0x7f8dc204d49a _PyEval_EvalFrameDefault + 1754
23      0x7f8dc204bfe2 /usr/local/python3/lib/libpython3.10.so.1.0(+0x22cfe2) [0x7f8dc204bfe2]
24      0x7f8dc1fef1ee /usr/local/python3/lib/libpython3.10.so.1.0(+0x1d01ee) [0x7f8dc1fef1ee]
25      0x7f8dc20f5fa6 /usr/local/python3/lib/libpython3.10.so.1.0(+0x2d6fa6) [0x7f8dc20f5fa6]
26      0x7f8dc20da7d4 /usr/local/python3/lib/libpython3.10.so.1.0(+0x2bb7d4) [0x7f8dc20da7d4]
27      0x7f8dc1c071ca /lib64/libpthread.so.0(+0x81ca) [0x7f8dc1c071ca]
28      0x7f8dc10d88d3 clone + 67
"
